core/sr-arithmetic/fuzzer/src/biguint.rs

```diff
@@ -0,0 +1,208 @@
+			});
+
+			run_with_data_set(4, &digits_u, &digits_v, |u, v| {
+				let expected = S::try_from(u.clone())
+					.unwrap()
+					.checked_sub(S::try_from(v.clone()).unwrap());
+				let t = u.clone().sub(&v);
+				if expected.is_none() {
+					assert!(t.is_err())
+				} else {
+					let t = t.unwrap();
+					let expected = expected.unwrap();
+					assert_eq!(
+						S::try_from(t.clone()).unwrap(), expected,
+						"{:?} - {:?} ===> {:?} != {:?}", u, v, t, expected,
+					);
+				}
+			});
+
+			run_with_data_set(2, &digits_u, &digits_v, |u, v| {
+				let expected = S::try_from(u.clone()).unwrap() * S::try_from(v.clone()).unwrap();
+				let t = u.clone().mul(&v);
+				assert_eq!(
+					S::try_from(t.clone()).unwrap(), expected,
+					"{:?} * {:?} ===> {:?} != {:?}", u, v, t, expected,
+				);
+			});
+
+			run_with_data_set(4, &digits_u, &digits_v, |u, v| {
+				let ue = S::try_from(u.clone()).unwrap();
+				let ve = S::try_from(v.clone()).unwrap();
+				if ve == 0 {
+					return;
+				}
+				let (q, r) = (ue / ve, ue % ve);
+				if let Some((qq, rr)) = u.clone().div(&v, true) {
+					assert_eq!(
+						S::try_from(qq.clone()).unwrap(), q,
+						"{:?} / {:?} ===> {:?} != {:?}", u, v, qq, q,
+					);
+					assert_eq!(
+						S::try_from(rr.clone()).unwrap(), r,
+						"{:?} % {:?} ===> {:?} != {:?}", u, v, rr, r,
+					);
+				} else if v.len() == 1 {
+					let qq = u.clone().div_unit(ve as Single);
+					assert_eq!(
+						S::try_from(qq.clone()).unwrap(), q,
+						"[single] {:?} / {:?} ===> {:?} != {:?}", u, v, qq, q,
+					);
+				} else if v.msb() != 0 && u.msb() != 0 && u.len() > v.len() {
+					panic!("div returned none for an unexpected reason");
+				}
+			});
+
+			// Test against num_bigint
+
+			let u = BigUint::from_limbs(&digits_u);
+			let v = BigUint::from_limbs(&digits_v);
+
+			digits_u.reverse();
+			digits_v.reverse();
+
+			let num_u = num_bigint::BigUint::new(digits_u.clone());
+			let num_v = num_bigint::BigUint::new(digits_v.clone());
+
+			// Equality
+
+			assert_biguints_eq(&u, &num_u);
+			assert_biguints_eq(&v, &num_v);
+
+			if digits_u == digits_v {
+				assert_eq!(u, v);
+				assert_eq!(num_u, num_v);
+			}
+
+			// Addition
+
+			let w = u.clone().add(&v);
+			let num_w = num_u.clone() + &num_v;
+
+			assert_biguints_eq(&w, &num_w);
+
+			// Subtraction
+
+			if let Ok(w) = u.clone().sub(&v) {
+				let num_w = num_u.clone() - &num_v;
+
+				assert_biguints_eq(&w, &num_w);
+			}
+
+			// Multiplication
+
+			let w = u.clone().mul(&v);
+			let num_w = num_u.clone() * &num_v;
+
+			assert_biguints_eq(&w, &num_w);
+
+			// Division
+
+			if v.len() == 1 && v.get(0) != 0 {
+				let w = u.clone().div_unit(v.get(0));
+				let num_w = num_u.clone() / &num_v;
+				assert_biguints_eq(&w, &num_w);
+			} else {
+				let w = u.clone().div(&v, rem).map(|(w, _)| w);
+				let num_w = num_u.clone().checked_div(&num_v);
+
+				// assert that the division was checked equally:
+				// assert_eq!(w.is_some(), num_w.is_some());
```

Comment:
and for this: strip both u and v at the very beginning. aJust better to do so. 
Then run and compare division only when u.len() > v.len()

In this case you can remote all the `if let` and unwrap both of them immediately, as they both _must_ return. then you can also either have `// assert_eq!(w.is_some(), num_w.is_some());` uncommented (though' I think unwrapping implies this)

Reply:
will this be solved or do you disagree?

Reply:
🏴